### PR TITLE
Added ts-jest

### PR DESCRIPTION
**Pull request by emin93**
In this PR I re-added `ts-jest` to enable out of the box jest testing with TypeScript.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "allowSyntheticDefaultImports": true,
     "esModuleInterop": true,
     "isolatedModules": true,
-    "jsx": "react-native",
+    "jsx": "react",
```
Comment (JReinhold) [inline tsconfig.json]: Just to get a deeper understanding, may I ask why this change is necessary? 🙂

Comment (emin93) [reply tsconfig.json]: I found out that `ts-jest` can't recognize JSX/TSX when this is set to `react-native`. You can easily reproduce it by switching it back to `react-native` and running `npm test` / `yarn test`, it will throw a `unexpected Token <`-error. I really appreciate your interest 😄

Comment (JReinhold) [reply tsconfig.json]: Oh great, I've had that exact problem before, I din't know it had anything to do with that setting. I'll have to try it out tomorrow!

**Comment by JReinhold** [comment]
This works great in my current project. Although I had to add an extra configuration to make it work with some of the more advanced libraries like `react-navigation` and `react-gesture-handler`. I'll leave the extra configuration here, in case anyone else comes by with the same problem:

Jest config (in `package.json` if using this template), gotten from https://github.com/kulshekhar/ts-jest/issues/604#issuecomment-413913615:
```json
...
"transformIgnorePatterns": ["/node_modules/(?!react-native)/.+"],
```

**Comment by emin93** [comment]
Oh that's very good to know! If I understood the regex correctly, it tells Jest to ignore everything that starts with `react-native` in the `node_modules` directory? Wouldn't it be better to ignore everything inside the `node_modules` directory?

**Comment by JReinhold** [comment]
It's actually the exact opposite, it tells jest to ignore all modules EXCEPT `react-native`-files. Notice the `!` at the beginning. I think jest defaults to ignore all `node_modules`, so this regex is basically whitelisting `react-native` in the transformation.

**Comment by emin93** [comment]
Sorry for my delayed answer, had quite a lot of action going on 😅

Ah I see, thanks for the explanation! I'm curious to know why ts-jest needs those files and if it makes sense to include that whitelisting config as a default in the template?

**Comment by JReinhold** [comment]
Yeah, I can't really tell for sure, why it needs that. I see that line in a lot of places and example repos, but can't find a specific answer to WHY its there.

IMHO it should be part of the default, because my experience is that you won't get very far without it. As you've experienced, you can test the classic hello world app without the transformIgnorePatterns, but anything beyond that will break in my experience.
